pyiceberg/table/__init__.py

```diff
@@ -1904,3 +1913,200 @@ def _generate_snapshot_id() -> int:
+            if name in partition_names:
+                raise ValueError(f"Cannot use partition name more than once {name}")
+            partition_names.add(name)
+
+        def _add_new_field(source_id: int, field_id: int, name: str, transform: Transform, partition_names: Set[int]):
+            _check_and_add_partition_name(name, source_id, partition_names)
+            return PartitionField(source_id, field_id, name, transform)
+            
+
+        for field in self._spec.fields:
+            if field.field_id not in self._deletes:
+                renamed = self._renames.get(field.name)
+                if renamed:
+                    new_field = _add_new_field(field.source_id, field.field_id, renamed, field.transform, partition_names)
+                else:
+                    new_field = _add_new_field(field.source_id, field.field_id, field.name, field.transform, partition_names)
+                last_assigned_field_id = max(last_assigned_field_id, new_field.field_id)
+                partition_fields.append(new_field)
+            elif self._table.format_version == 1:
+                renamed = self._renames.get(field.name)
+                if renamed:
+                    new_field = _add_new_field(field.source_id, field.field_id, renamed, VoidTransform(), partition_names)
+                else:
+                    new_field = _add_new_field(field.source_id, field.field_id, field.name, VoidTransform(), partition_names)
+
+                last_assigned_field_id = max(last_assigned_field_id, new_field.field_id)
+                partition_fields.append(new_field)
+        
+        for added_field in self._adds:
+            new_field = PartitionField(source_id=added_field.source_id, field_id=added_field.field_id, transform=added_field.transform, name=added_field.name)
+            partition_fields.append(new_field)
+        # Reuse spec id or create a new one.
+        new_spec = PartitionSpec(*partition_fields)
+        new_spec_id = INITIAL_PARTITION_SPEC_ID
+        for spec in self._table.specs().values():
+            if new_spec.compatible_with(spec):
+                new_spec_id = spec.spec_id
+                break
+            elif new_spec_id <= spec.spec_id:
+                new_spec_id = spec.spec_id + 1
+
+        return PartitionSpec(*partition_fields, spec_id=new_spec_id)
+    
+    def _check_redundant_partitions(self, field: PartitionField):
+        if isinstance(field.transform, TimeTransform):
+            existing_time_field = self._added_time_fields.get(field.source_id)
+            if existing_time_field:
+                raise ValueError(f"Cannot add time partition field: {field.name} conflicts with {existing_time_field.name}")
+            self._added_time_fields[field.source_id] = field
+    
+    def _partition_field(self, transform_key: Tuple[int, Transform], name: str) -> PartitionField:
+        if self._table.metadata.format_version == 2 and self._table.metadata:
```

Comment:
I don't think metadata can be `None`:
```suggestion
        if self._table.metadata.format_version == 2:
```